java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -124,58 +124,53 @@ private void run(@Nullable final ExplodedGraph.Node currentNode, @Nullable final
-    return arguments.indexOf(candidate);
-  }
-
-  private static List<SymbolicValue> argumentsUsedForMethodInvocation(ExplodedGraph.Node invocationNode, MethodInvocationTree mit) {
-    List<SymbolicValue> values = new ArrayList<>(invocationNode.programState.peekValues(mit.arguments().size()));
-    Collections.reverse(values);
-    return values;
-  }
-
-  private static String learnedConstraintMessage(Constraint lc, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
+  private static Stream<JavaFileScannerContext.Location> learnedConstraintFlow(Constraint learnedConstraint, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
     Tree nodeTree = parent.programPoint.syntaxTree();
-    String name = SyntaxTreeNameFinder.getName(nodeTree);
     if (nodeTree.is(Tree.Kind.METHOD_INVOCATION)) {
-      MethodYield selectedMethodYield = currentNode.selectedMethodYield(parent);
-      if (selectedMethodYield != null) {
-        Type exceptionType = selectedMethodYield.exceptionType();
-        // Only considering exceptional flows
-        if (exceptionType != null) {
-          return exceptionType.isUnknown() ? String.format("Exception thrown by '%s'", name) : String.format("Exception '%s' thrown by '%s'", exceptionType.name(), name);
-        }
-      }
-      return String.format("'%s' returns %s", name, lc.valueAsString());
+      return methodInvocationFlow(learnedConstraint, currentNode, parent);
     }
     if (nodeTree.is(Tree.Kind.NEW_CLASS)) {
-      return String.format("Constructor call creates '%s'", lc.valueAsString());
+      return Stream.of(location(parent, String.format("Constructor call creates '%s'", learnedConstraint.valueAsString())));
+    }
+    String name = SyntaxTreeNameFinder.getName(nodeTree);
+    String message = name == null ? learnedConstraint.valueAsString() : String.format("Implies '%s' is %s", name, learnedConstraint.valueAsString());
+    return Stream.of(location(parent, message));
+  }
+
+  private static Stream<JavaFileScannerContext.Location> methodInvocationFlow(Constraint learnedConstraint, ExplodedGraph.Node currentNode, ExplodedGraph.Node parent) {
+    MethodInvocationTree mit = (MethodInvocationTree) parent.programPoint.syntaxTree();
+    MethodYield selectedMethodYield = currentNode.selectedMethodYield(parent);
+    if (selectedMethodYield == null) {
+      return Stream.of(location(parent, String.format("'%s' invoked", mit.symbol().name())));
+    }
+    Stream.Builder<JavaFileScannerContext.Location> flow = Stream.builder();
+    if (selectedMethodYield.isExceptional()) {
+      Type exceptionType = selectedMethodYield.exceptionType();
+      String exception = exceptionType == null || exceptionType.isUnknown() ? "Exception" : exceptionType.name();
+      flow.add(location(parent, String.format("'%s' thrown by '%s'", exception, mit.symbol().name())));
+    }
+    if (selectedMethodYield.resultConstraint == learnedConstraint) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)